### ISSUE TITLE
Collectd seems to be processing packets out of order and triggering "Value too old"

### ISSUE THREAD

**Issue by zeridon** [collectd/collectd#987]
Currently we are experiencing the "Value Too old" error on a regular basis.

The error looks like this:

    uc_update: Value too old: name = i-<redacted>/<redacted>/gauge-min; value time = 1427467785.000; last cache update = 1427467805.000;

The setup we have is:
 
 * Java App (using jcollectd) -> collectd
 * JCollectd bundles several updates (1k minimum) in our case this are 9 data points
 * Each packet dispatched to collectd contains 9 data points

When it seems data points from one packet are not strictly ordered by timestamp at all times and sometimes a data point with later timestamp is inserted before an earlier one is (i could of course be wrong).

Sample tcpdump is available on request only.

**Comment by rubenk** [comment]
Hi!
Does this go away when you set WriteThreads 1? What write plugin are you using?

**Comment by zeridon** [comment]
Setting the ```WriteThreads``` to 1 seems to do the trick (will report after running like this for some time).

As for the write plugins i have only ```write_graphite``` enabled (besides syslog)

**Comment by zeridon** [comment]
This fix seems to have corrected the issue.

Can we convert this to a request that data points for a metric should be handled by a single thread only to avoid out of order writing of the data points (and therefore rejecting them in some cases).

**Comment by rubenk** [comment]
There is no specific request for that, but there are multiple issues that report this already, for instance #450

**Comment by octo** [comment]
There is a problem when multiple updates of the same metric are in one network packet because they get "dispatched" effectively simultaneously. Multiple write threads will try to write consecutive values for one metric concurrently, which may or may not work.

I think it would make sense to "fix" this on the sending side and check in the Java code if a metric is already staged and send out the buffer if so.

**Comment by zeridon** [comment]
Yes this indeed seems to be the issue. The packets (comming from the java lib) in order to optimize network traffic were bundling several rounds of data. This as you say leads to issues with multiple write threads.

This currently got solved by lowering the write threads to just 2 (or one, but can't check at the moment). The java library being used is ... very resistant to lowering the packet size (and also we like the savings).

At some point i was pondering on the idea to have some internal ordered cache in order to be able to bunch updates, but this got nowhere as my c/c++ skills are not so good.

**Comment by octo** [comment]
This is very hard to fix, even with C skills. I'll talk to the other *collectd* team members and see if we can come up with a solution for this.

**Comment by hulbs9nw** [comment]
Is there any update on this?  Would an alternative workaround be to decrease the MaxPacketSize but increase the number of write threads?

**Comment by zeridon** [comment]
@hulbs9nw the problem is that if you have more than one value for a metric in the same received packet those can be dispatched to different write threads (and those do not sync) so older value can be writen out before an earlier value. IF (and that is a big if) you can assure that there is not more than one data point per metric per packet you are safe. If not you must lower write threads to 1. Packet size can work for reducing the number of datapoints per metric per packet but this might lead to suboptimal network performance.

**Comment by hulbs9nw** [comment]
Thanks for the info zeridon, that is how I understood the problem.  Even if I lower the MaxPacketSize it's not guaranteed that it will contain only a single metric.  The reason I ask is our collectd server is busy and it looks like we're being limited by the single write thread.

Is this "bug" specific to the rrdcached plugin or will it affect any plugin that utilises multiple write threads?  Another option for us would be to ditch rrdcached and look to move to something like influxdb.

![screenshot from 2018-02-12 09-28-33](https://user-images.githubusercontent.com/8561002/36090203-64d6ffbc-0fd7-11e8-89cb-a78d75ed5aeb.png)

**Comment by pforemski** [comment]
I'm experiencing the bug when using the curl_json plugin in collectd 5.7.0 (from jessie-backports).

Setting "WriteThreads 1" does NOT help. I can see UDP packets generated by the network plugin being occasionally sent each 20s (instead of 10s), which triggers the log message.

IMHO, this is quite an important bug that should be well documented.

**Comment by cbz** [comment]
I'm also seeing this when sending statistics from a remote collectd. Setting WriteThreads 1 doesn't appear to help in this case either.

**Comment by eero-t** [comment]
Note: this message triggers also if plugin reports on same plugin read call (e.g. by accident / due to bug) a metric with exactly the same name multiple times.
